src/Strings.h

```diff
@@ -0,0 +1,388 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Details {
+
+template<typename IteratorT>
+inline std::vector<IteratorT> IteratorVectorGenerator(IteratorT const & begin,
+                                                     IteratorT const & end) {
+    std::vector<IteratorT> wordIterVector;
+    wordIterVector.reserve(static_cast<unsigned int>(std::distance(begin, end) + 1));
+
+    for (IteratorT strIter = begin; strIter != end; ++strIter)
+        wordIterVector.emplace_back(std::move(strIter));
+    wordIterVector.emplace_back(std::move(end));
+
+    return wordIterVector;
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void Parse(IteratorT const &begin,
+                  IteratorT const &end,
+                  IsDelimiterT const &isDelimiter,
+                  std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    IteratorT left = begin;
+    IteratorT right = left;
+    while (left != end && *left == isDelimiter)
+        ++left;
+
+    right = left;
+    while (right != end) {
+        if (*right == isDelimiter) {
+            callback(left, right);
```

Comment:
>callback(left, right); [](start = 12, length = 22)

According to the signature, the callback will modify the iterators but it is not a legal thing to do here since they are used for iteration in this function.